parquet-column/src/test/java/org/apache/parquet/column/values/bytestreamsplit/ByteStreamSplitValuesEndToEndTest.java

```diff
@@ -0,0 +1,106 @@
+/* 
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.column.values.bytestreamsplit;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.Random;
+
+import org.apache.parquet.bytes.ByteBufferInputStream;
+import org.apache.parquet.bytes.BytesInput;
+import org.apache.parquet.bytes.DirectByteBufferAllocator;
+import org.junit.Test;
+
+public class ByteStreamSplitValuesEndToEndTest {
+
+  @Test
+  public void TestFloatPipeline() throws Exception {
+    // Generate random data.
+    Random rand = new Random(1337);
+    final int numElements = 1024;
+    float[] values = new float[numElements];
+    for (int i = 0; i < numElements; ++i) {
+      float f = rand.nextFloat() * 4096.0f;
+      values[i] = f;
+    }
+
+    // Encode data.
+    ByteStreamSplitValuesWriter.FloatByteStreamSplitValuesWriter writer
+            = new ByteStreamSplitValuesWriter.FloatByteStreamSplitValuesWriter(
+                    numElements * 4, numElements * 4, new DirectByteBufferAllocator());
+    for (float v : values) {
+      writer.writeFloat(v);
+    }
+
+    assertEquals(numElements * 4, writer.getBufferedSize());
+    BytesInput input = writer.getBytes();
+    assertEquals(numElements * 4, input.size());
+
+    ByteStreamSplitValuesReaderForFloat reader
+            = new ByteStreamSplitValuesReaderForFloat();
+
+    reader.initFromPage(numElements, ByteBufferInputStream.wrap(input.toByteBuffer()));
+
+    for (float expectedValue : values) {
+      float newValue = reader.readFloat();
+      assertEquals(expectedValue, newValue, 0.0f);
+    }
+
+    writer.reset();
+    writer.close();
```

Comment:
I would suggest invoking reset/close in the `finally` block after putting the whole code in a `try` block. This way the writer will be reset and closed even if an exception is occurred during the execution.

Reply:
I addressed this in my patch. I was thinking that if an exception is thrown, then this should be still fine since the garbage collector would destroy the writer and the stream would be closed.

Reply:
As per the java spec it is not guaranteed that the method `finalize` is invoked when the object is garbage collected. So, one shall ensure that the resources occupied outside of the JVM (e.g. I/O streams) are freed at some point (before the references are lost).